src/v/kafka/server/request_context.h

```diff
@@ -275,6 +267,33 @@ class request_context {
         usage_mgr().add_bytes_sent(response_size - internal_bytes_sent);
     }
 
+    template<typename ResponseType>
+    ss::future<> log_response(ResponseType res) {
+        // Create the log message early because members _conn and _header are
+        // destroyed at the following co_await
+        auto msg = fmt::format(
+          "[{}:{}] sending {}:{} for {}, response {}",
+          _conn->client_host(),
+          _conn->client_port(),
+          ResponseType::api_type::key,
+          ResponseType::api_type::name,
+          _header.client_id,
+          res);
+
+        co_await ss::do_with(
+          std::move(res), std::move(msg), [](auto& res, auto& msg) {
+              return res.data.errored().then([&msg](auto errored) {
+                  if (errored) {
+                      thread_local static ss::logger::rate_limit rate(
```

Comment:
I feel like logging errors is more interesting than logging successes, but now we only log up to one error every 10s?

Reply:
Yeah I think we want logic like this maybe? (Psuedo code)

```
if (kafka is at trace logging) {
  log_everything(response)
} else if (kafka is at debug logging) {
  log_rate_limited_errors(response)
}
```

Reply:
> now we only log up to one error every 10s?

Yes. The idea is that a broker could report many errors. This rate limit is so the logs are not flooded. I will send you the slack threads for context

Reply:
> Yeah I think we want logic like this maybe? (Psuedo code)
> 
> ```
> if (kafka is at trace logging) {
>   log_everything(response)
> } else if (kafka is at debug logging) {
>   log_rate_limited_errors(response)
> }
> ```

Ah I missed that in our slack conversations. I'll add that into the next push.

Reply:
> > now we only log up to one error every 10s?
> 
> Yes. The idea is that a broker could report many errors. This rate limit is so the logs are not flooded. I will send you the slack threads for context

I don't think we want to eliminate the logging of errors at trace level.

Would something like this work?
* Leave Trace logging for responses
* Log a sample of errors at debug

I wonder if the reate limiter can also be queried along with the log_level to reduce the amount of work that is done when there is a high rate of errors but that will be limited

Reply:
> Would something like this work?
> * Leave Trace logging for responses
> * Log a sample of errors at debug

I still think it would probably be ok to just log all errors at debug,

The current problem is that the only way to get the errors is to turn on trace logging, which is very noisy. Elevating errors to debug seems reasonanle.

At the moment the code looks like it throws away almost all errors, which seems counterproductive to increasing the visibility of errors.